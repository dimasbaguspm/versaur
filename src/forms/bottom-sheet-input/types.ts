import type { TextInputProps } from '../text-input/types'
import type { ChangeEventHandler, InputHTMLAttributes, ReactNode } from 'react'

export interface BottomSheetInputContextValue {
  isOpen: boolean
  setIsOpen: (open: boolean) => void
  onChange?: ChangeEventHandler<HTMLInputElement>
  value: InputHTMLAttributes<HTMLInputElement>['value']
}

/**
 * BottomSheetInputProps extends TextInputProps, but input is always readOnly
 * @property {boolean} open - BottomSheet open state (controlled)
 * @property {(open: boolean) => void} onOpenChange - Callback for sheet open state change
 * @property {React.ReactNode} children - BottomSheet content
 */
export interface BottomSheetInputProps
  extends Omit<TextInputProps, 'readOnly' | 'children'> {
  children: (ctx: BottomSheetInputContextValue) => ReactNode
}
