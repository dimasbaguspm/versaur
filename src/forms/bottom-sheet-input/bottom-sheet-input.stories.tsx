import { BottomSheetInput } from './bottom-sheet-input'
import { useState } from 'react'
import { Button } from '@/primitive'
import { SearchInput } from '../search-input'

export default {
  title: 'Forms/BottomSheetInput',
  component: BottomSheetInput,
}

export const Basic = () => {
  const [value, setValue] = useState('')

  return (
    <BottomSheetInput
      label='Pick a value'
      placeholder='Click to open bottom sheet'
      value={value}
      onChange={e => setValue(e.target.value)}
    >
      {ctx => (
        <>
          <BottomSheetInput.Body>
            <SearchInput
              autoFocus
              label='Search Transactions'
              value={value}
              onChange={ctx.onChange}
              placeholder='Type something...'
            />
          </BottomSheetInput.Body>
          <BottomSheetInput.Footer>
            <Button
              onClick={() => ctx.setIsOpen(false)}
              className='btn btn-primary'
            >
              Confirm
            </Button>
          </BottomSheetInput.Footer>
        </>
      )}
    </BottomSheetInput>
  )
}
