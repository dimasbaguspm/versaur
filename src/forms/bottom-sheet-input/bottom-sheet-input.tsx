import { useState, useMemo, forwardRef } from 'react'
import { TextInput } from '../text-input/text-input'
import { BottomSheet } from '@/overlays/bottom-sheet/bottom-sheet'
import type {
  BottomSheetInputProps,
  BottomSheetInputContextValue,
} from './types'

/**
 * BottomSheetInput is a readOnly TextInput that opens a BottomSheet when clicked.
 * BottomSheet content is provided by consumer as children.
 */
const BottomSheetInputRoot = forwardRef<
  HTMLInputElement,
  BottomSheetInputProps
>(({ children, onChange, value, ...textInputProps }, ref) => {
  const [isOpen, setIsOpen] = useState(false)

  const contextValue = useMemo(
    () => ({ isOpen, setIsOpen, onChange, value }),
    [isOpen, setIsOpen, onChange, value]
  ) satisfies BottomSheetInputContextValue

  return (
    <>
      <TextInput
        {...textInputProps}
        onChange={onChange}
        value={value}
        ref={ref}
        readOnly
        tabIndex={0}
        onClick={() => setIsOpen(true)}
        aria-haspopup='dialog'
        aria-expanded={isOpen}
        onKeyDown={e => {
          if (e.key === 'Enter' || e.key === ' ') {
            e.preventDefault()
            setIsOpen(true)
          }
        }}
      />
      <BottomSheet isOpen={isOpen} onClose={() => setIsOpen(false)}>
        {isOpen && children(contextValue)}
      </BottomSheet>
    </>
  )
})

export const BottomSheetInput = Object.assign(BottomSheetInputRoot, {
  Header: BottomSheet.Header,
  Body: BottomSheet.Body,
  Footer: BottomSheet.Footer,
  Title: BottomSheet.Title,
})
