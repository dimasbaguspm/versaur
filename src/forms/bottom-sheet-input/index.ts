export { BottomSheetInput } from './bottom-sheet-input'
export type * from './types'
