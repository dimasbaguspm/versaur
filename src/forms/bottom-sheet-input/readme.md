# BottomSheetInput

A mobile-friendly input component that opens a BottomSheet overlay for custom content selection. Follows the same compound/context pattern as DrawerInput, but uses BottomSheet for mobile overlays.

## Usage

```tsx
import { BottomSheetInput } from '@/forms/bottom-sheet-input'

const [value, setValue] = useState('')

<BottomSheetInput
  label='Pick a value'
  placeholder='Click to open bottom sheet'
  value={value}
  onChange={e => setValue(e.target.value)}
>
  {ctx => (
    <>
      <BottomSheetInput.Header>
        <Text fontSize='lg' fontWeight='bold'>Bottom Sheet Input Example</Text>
      </BottomSheetInput.Header>
      <BottomSheetInput.Body>
        <SearchInput
          label='Search Transactions'
          value={value}
          onChange={ctx.onChange}
          placeholder='Type something...'
        />
      </BottomSheetInput.Body>
      <BottomSheetInput.Footer>
        <Button onClick={() => ctx.setIsOpen(false)} className='btn btn-primary'>Confirm</Button>
      </BottomSheetInput.Footer>
    </>
  )}
</BottomSheetInput>
```

## API

- All TextInput props except `readOnly` and `children`
- Compound pattern: `Header`, `Body`, `Footer`, `Title`
- Context: `{ isOpen, setIsOpen, onChange, value }`

## Accessibility

- Uses `aria-haspopup='dialog'` and `aria-expanded` on input
- Follows WCAG 2.1 AA standards

## Testing

- Colocate tests in `__tests__/bottom-sheet-input.test.tsx`
- Use Vitest + React Testing Library
- Assert HTML output and accessibility
