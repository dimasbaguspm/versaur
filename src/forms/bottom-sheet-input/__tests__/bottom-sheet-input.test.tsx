import { render, screen, fireEvent } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../bottom-sheet-input.stories'

describe('BottomSheetInput', () => {
  const { Basic } = composeStories(stories)

  it('renders correctly (snapshot)', () => {
    const { asFragment } = render(<Basic />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('opens bottom sheet on input click', () => {
    render(<Basic />)
    const input = screen.getByPlaceholderText('Click to open bottom sheet')
    fireEvent.click(input)
    expect(screen.getByRole('searchbox')).toBeInTheDocument()
  })

  it('closes bottom sheet on confirm', () => {
    render(<Basic />)
    const input = screen.getByPlaceholderText('Click to open bottom sheet')
    fireEvent.click(input)
    const confirmBtn = screen.getByText('Confirm')
    fireEvent.click(confirmBtn)
    expect(confirmBtn).not.toBeVisible()
  })
})
