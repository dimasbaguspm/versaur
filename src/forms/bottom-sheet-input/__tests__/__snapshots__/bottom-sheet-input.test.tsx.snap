// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BottomSheetInput > renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block text-sm font-medium text-foreground mb-2"
      for="«r0»"
    >
      Pick a value
    </label>
    <div
      class="relative w-full"
    >
      <input
        aria-expanded="false"
        aria-haspopup="dialog"
        aria-invalid="false"
        class="block w-full rounded-md border bg-white transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none disabled:bg-gray-50 border-primary/30 text-foreground focus:border-primary focus:ring-primary/20 pl-3 pr-3 h-10"
        id="«r0»"
        placeholder="Click to open bottom sheet"
        readonly=""
        tabindex="0"
        value=""
      />
    </div>
  </div>
  <div
    aria-hidden="true"
    class="fixed inset-0 z-40 duration-300 pointer-events-none"
  />
  <div
    aria-modal="true"
    class="fixed left-0 bottom-0 z-40 w-full max-h-[90dvh] bg-background shadow-lg border-t border-border rounded-lg transition-transform duration-300 ease-in-out will-change-transform translate-y-full"
    role="dialog"
    tabindex="-1"
  />
</DocumentFragment>
`;
